helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -0,0 +1,354 @@
+  }
+
+  private static class ClusterState {
+    List<InstanceConfig> instanceConfigs = new ArrayList<>();
+    ClusterConfig clusterConfig;
+    List<String> resources = new ArrayList<>();
+    List<String> instances;        // cluster LiveInstance + addInstances - instancesToRemove.
+  }
+
+  // Result format: Map of resource -> partition -> instance -> state.
+  private static class AssignmentResult extends HashMap<String, Map<String, Map<String, String>>> {
+    public AssignmentResult() {
+      super();
+    }
+  }
+
+  public static final String INSTANCE_CHANGE = "InstanceChange";
+  public static final String INSTANCE_CHANGE_ADD_INSTANCES = "AddInstances";
+  public static final String INSTANCE_CHANGE_REMOVE_INSTANCES = "RemoveInstances";
+  public static final String INSTANCE_CHANGE_SWAP_INSTANCES = "SwapInstances";
+  public static final String OPTIONS = "Options";
+  public static final String OPTIONS_INSTANCE_FLT = "InstanceFilter";
+  public static final String OPTIONS_RESOURCE_FLT = "ResourceFilter";
+
+  @ResponseMetered(name = HttpConstants.WRITE_REQUEST)
+  @Timed(name = HttpConstants.WRITE_REQUEST)
+  @POST
+  public Response computePotentialAssignment(@PathParam("clusterId") String clusterId,
+      String content) {
+
+    InputFields inputFields;
+    ClusterState clusterState;
+    AssignmentResult result;
+
+    try {
+      // 1.  Try to parse the content string. If parseable, use it as a KV map. Otherwise, return a REASON String
+      inputFields = readInput(content);
+      // 2. Read cluster status from ZK.
+      clusterState = readClusterStateAndValidateInput(clusterId, inputFields);
+      // 3. Call rebalancer tools for each resource.
+      result = computeOptimalAssignmentForResources(inputFields, clusterState, clusterId);
+      // 4. Serialize result to JSON and return.
+      return JSONRepresentation(result);
+    } catch (InvalidParameterException ex) {
+      return badRequest(ex.getMessage());
+    } catch (JsonProcessingException e) {
+      return badRequest("Invalid input: Input can not be parsed into a KV map." + e.getMessage());
+    } catch (OutOfMemoryError e) {
+      LOG.error("OutOfMemoryError while calling partitionAssignment" + Arrays
+          .toString(e.getStackTrace()));
+      return badRequest(
+          "Response size is too large to serialize. Please query by resources or instance filter");
+    } catch (Exception e) {
+      LOG.error("Failed to compute partition assignment:" + Arrays.toString(e.getStackTrace()));
+      return badRequest("Failed to compute partition assignment: " + e);
+    }
+  }
+
+  private InputFields readInput(String content)
+      throws InvalidParameterException, JsonProcessingException {
+
+    Map<String, Map<String, Object>> customFieldsMap;
+    InputFields inputMap = new InputFields();
+    customFieldsMap =
+        OBJECT_MAPPER.readValue(content, new TypeReference<HashMap<String, Map<String, Object>>>() {
```

Comment:
If you define a class with the field names like INSTANCE_CHANGE, then ObjectMapper can handle it automatically for you. It would help to simplify code, and more importantly, reduce bug and maintenance costs.

Reply:
TFTR. Updated.